frappe/deferred_insert.py

```diff
@@ -27,20 +27,19 @@ def deferred_insert(doctype: str, records: list[Union[dict, "Document"]] | str):
 
 def save_to_db():
 	queue_keys = frappe.cache.get_keys(queue_prefix)
+	# TODO: can this be async?
 	for key in queue_keys:
-		record_count = 0
 		queue_key = get_key_name(key)
 		doctype = get_doctype_name(key)
-		while frappe.cache.llen(queue_key) > 0 and record_count <= 500:
-			records = frappe.cache.lpop(queue_key)
+		if cache_size := frappe.cache.llen(queue_key):
+			records = frappe.cache.lpop(queue_key, count=cache_size)
```

Comment:
- I don't think popping the entire key is a good idea here - In case of *BaseException* (any unhandled exception) in the following steps, all data will be lost
- About the added TODO: bulk_insert would not trigger events + a failure in single doc's data insertion can lead to full failure

Reply:
> * I don't think popping the entire key is a good idea here - In case of _BaseException_ (any unhandled exception) in the following steps, all data will be lost

in that case, `lpop` makes sense after operation has completed while in this step `lrange` can give access to all items. it can be refactored to idiomatically handle in a single context manager where exceptions and success cases are automatically managed. in case of exception, the rollback is a no-op. but in case of any failure, we could allow for a retry cap. for simplicity, what would be the ideal failover strategy? a retry cap or skip?

> About the added TODO: bulk_insert would not trigger events + a failure in single doc's data insertion can lead to full failure

I am not sure the chain of commands this function follows and impact surface area so I'll drop it. it could be a different feature request if this impacts performance anyway.

Reply:
- In contrast to previous limit, this change removes the limit with bulk pop while retrying for failed entries with the assumption that the records are valid and would eventually go through
- `insert_record` is kept simple and generic exception refactored out.